File: glue_and_networking/checksum.c
#include "checksum.h"
#include "packet.h"

uint16_t ones_checksum (char* packet_bytes, int len)
{
	uint16_t checksum = 0;
	uint16_t current_word;
	int i = 0;

	while (i < len)
	{
		current_word = packet_bytes[i] << 8;
		i++;
		if (i < len)
			current_word |= packet_bytes[i];
		i++;

		checksum += current_word;
		if ((checksum >> 16) == 1)
		{
			checksum = checksum & 0xFFFF;
			checksum++;
		}
	}

	checksum = ~checksum;

	return checksum;
}

char check_checksum (char* packet_bytes, int len);
{
	struct packet_hdr* header = (struct packet_hdr*)packet_bytes;
	uint16_t transmitted_checksum = header->checksum;

	header->checksum = 0;

	return ones_checksum(packet_bytes) == transmitted_checksum;
}

File: glue_and_networking/checksum.py
def ones_checksum(packet_bytes):
	checksum = 0

	if len(packet_bytes) % 2 == 1
		packet_bytes = packet_bytes + bytearray(1)

	for i in range(0, len(packet_bytes)/2)
		current_word = packet_bytes[i*2] << 8
		current_word |= packet_bytes[i*2 + 1]

		checksum += current_word
		if (checksum >> 16) == 1
			checksum = checksum & 0xFFFF
			checksum += 1

	checksum = ~checksum

	return checksum

def check_checksum(packet_bytes):
	transmitted_checksum = packet_bytes[0] << 8
	transmitted_checksum |= packet_bytes[1]

	packet_bytes[0] = 0
	packet_bytes[1] = 0

	return ones_checksum(packet_bytes) == transmitted_checksum
